### DISCUSSION TITLE
Is there any proper way to handle validation exceptions in @OnOpen method?

### DISCUSSION THREAD

**Discussion by domi-nika60** [micronaut-projects/micronaut-core#5320] [Q&A]
Hi,

Just wondering if there is any way to handle exceptions in Websocket server onOpen method.
In my project I introduced validation of the parameters being sent via websocket connection. 

Something like this:
```
@OnOpen
@Validated
public Publisher<String> onOpen(
        @Valid @RequestBean WebsocketClient data,
        WebSocketSession session
) {
...
}
```
And in WebsocketClient class:
```
@NotBlank
@PathVariable
@Pattern(regexp = "<some regexp>")
private String id;
```
When the `id` is not valid, there is an exception on OnOpen method with all stacktrace. I was trying to handle it in any way. Most reasonable was to add:
(from [https://guides.micronaut.io/micronaut-error-handling/guide/index.html](https://guides.micronaut.io/micronaut-error-handling/guide/index.html ))
```
@OnOpen
@Validated
@Error(exception = ConstraintViolationException.class)
  public Publisher<String> onOpen(
        @Valid @RequestBean WebsocketClient data,
        WebSocketSession session,
        ConstraintViolationException ex
) 
```
And use this execption to be logged.
But then there is an error 

```
Parameter to @OnOpen must either be a URI variable, a WebSocketSession , the HttpRequest, or annotated with an HTTP binding annotation (such as @Header)
        ConstraintViolationException ex
                                     ^
```
                                     
 Is there any other proper way to handle validation exceptions in @OnOpen methods?
Maybe it is a bug?

**Comment by graemerocher** [comment]
You would need to add an `@OnError` handler for the WebSocket

**Reply by domi-nika60** [reply]
Can you show me how to do it? Is there any documentation somewhere?

**Reply by graemerocher** [reply]
there is documentation here describing the annotation https://docs.micronaut.io/latest/guide/#websocketServer

Though no example granted

**Reply by domi-nika60** [reply]
Does this `@OnError` is addressed to `@Valid` annotation of WebSocket data? For me it seems to not work like this.

**Reply by graemerocher** [reply]
Should do, if not provide an example in an issue. Here is an example from our test suite https://github.com/micronaut-projects/micronaut-core/blob/b52ef797a2b4f5429cd38768b5a3d2534c717c8c/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/websocket/errors/MessageErrorSocketWithOnError.java

**Reply by domi-nika60** [reply]
I've made sth like this:

```
    @OnOpen
    @Validated
    public Publisher<String> onOpen(
        @Valid @RequestBean WebsocketClient data,
        WebSocketSession session
    ) {...}
    
    @OnError
    public void onError(ConstraintViolationException e) {
        LOG.error("Error occured: {}", e );
    }
 ```
 
 But the `OnError` method is not being executed. There is no difference in application behavior and logs if `OnError` is there or not. Any idea why? What do I do wrong?

**Reply by domi-nika60** [reply]
Any updates about this? If not, I will rise an issue.

**Reply by graemerocher** [reply]
Seems issue report worthy
